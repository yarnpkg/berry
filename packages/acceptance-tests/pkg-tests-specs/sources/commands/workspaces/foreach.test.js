const {
  fs: {writeJson, writeFile},
} = require('pkg-tests-core');

async function setupWorkspaces(path) {
  await writeFile(`${path}/mutexes/workspace-a`, ``);
  await writeFile(`${path}/mutexes/workspace-b`, ``);

  await writeFile(`${path}/.yarnrc`, `plugins:\n  - ${JSON.stringify(require.resolve(`@berry/monorepo/scripts/plugin-workspace-tools.js`))}\n`);

  await writeFile(`${path}/packages/workspace-a/server.js`, getClientContent(`${path}/mutexes/workspace-a`, `PING`));
  await writeJson(`${path}/packages/workspace-a/package.json`, {
    name: `workspace-a`,
    version: `1.0.0`,
    scripts: {
      print: `echo Test Workspace A`,
      start: `node server.js`,
    }
  });

  await writeFile(`${path}/packages/workspace-b/client.js`, getClientContent(`${path}/mutexes/workspace-b`, `PONG`));
  await writeJson(`${path}/packages/workspace-b/package.json`, {
    name: `workspace-b`,
    version: `1.0.0`,
    scripts: {
      print: `echo Test Workspace B`,
      start: `node client.js`
    },
    dependencies: {
      [`workspace-a`]: `workspace:*`,
      [`workspace-c`]: `workspace:*`
    }
  });

  await writeJson(`${path}/packages/workspace-c/package.json`, {
    name: `workspace-c`,
    version: `1.0.0`,
    scripts: {
      print: `echo Test Workspace C`,
    },
    dependencies: {
      [`workspace-a`]: `workspace:*`
    }
  });
}

describe(`Commands`, () => {
  describe(`workspace foreach`, () => {
    test(
      `runs scripts in parallel with interlaced report`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspaces`, `foreach`, `run`, `--parallel`, `--interlaced`, `--jobs=2`, `start`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          const lines = stdout.trim().split(`\n`);
          const firstLine = lines[0];

          let isInterlaced = false;

          // Expect Done on the last line
          expect(lines.pop()).toContain(`Done`);
          expect(lines.length).toBeGreaterThan(0);
          expect(code).toBe(0);
          expect(stderr).toBe(``);

          for (let i = 1; i < lines.length / 2; i++) {
            if (firstLine !== lines[i])
              isInterlaced = true;
          }

          expect(isInterlaced).toBe(true);
        }
      )
    );

    test(
      `runs scripts in parallel in topological order`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspaces`, `foreach`, `run`, `--parallel`, `--with-dependencies`, `--jobs=2`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );

    test(
      `runs scripts topological order by default`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspaces`, `foreach`, `run`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );

    test(
      `shows prefix with flag`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspaces`, `foreach`, `run`, `--prefixed`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );

    test(
      `run script only on included workspaces using [-i,--include]`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspaces`, `foreach`, `run`, `--prefixed`, `--include`, `workspace-a`, `-i`, `workspace-b`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );

    test(
      `excludes workspaces from running scripts using [-x,--exclude]`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspaces`, `foreach`, `run`, `--prefixed`, `--exclude`, `workspace-a`, `-x`, `workspace-b`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );

    test(
      `throws an error if using jobs without parallel`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          await run(`install`);
          await expect(run(`workspaces`, `foreach`, `run`, `--jobs=2`, `print`)).rejects.toThrowError(/parallel must be set/);
        }
      )
    );

    test(
      `throws an error if jobs is lower than 2`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`packages/*`]
        },
        async ({ path, run }) => {
          await setupWorkspaces(path);

          await run(`install`);
          await expect(run(`workspaces`, `foreach`, `run`, `--parallel`, `--jobs=1`, `print`)).rejects.toThrowError(/jobs must be greater/);
        }
      )
    );
  });
});

function getClientContent(mutex, name) {
  return `
    const fs = require('fs');
    const path = require('path');

    const mutex = ${JSON.stringify(mutex)};
    const mutexDir = path.dirname(mutex);

    async function handshake() {
      fs.unlinkSync(mutex);

      while (fs.readdirSync(mutexDir).length !== 0) {
        await sleep();
      }
    }

    async function sleep() {
      await new Promise(resolve => {
        setTimeout(resolve, 16);
      });
    }

    async function main() {
      await handshake();

      for (let t = 0; t < 60; ++t) {
        process.stdout.write(${JSON.stringify(name)} + '\\n');
        await sleep();
      }
    }

    main();
  `;
}
