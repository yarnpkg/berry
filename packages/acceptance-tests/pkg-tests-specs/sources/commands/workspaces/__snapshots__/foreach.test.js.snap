// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands workspace foreach excludes workspaces from running scripts using [-x,--exclude] 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: [workspace-c]: Test Workspace C
➤ YN0000: Done
",
}
`;

exports[`Commands workspace foreach run script only on included workspaces using [-i,--include] 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: [workspace-a]: Test Workspace A
➤ YN0000: [workspace-b]: Test Workspace B
➤ YN0000: Done
",
}
`;

exports[`Commands workspace foreach runs scripts in parallel in topological order 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Test Workspace A
➤ YN0000: Test Workspace C
➤ YN0000: Test Workspace B
➤ YN0000: Done
",
}
`;

exports[`Commands workspace foreach runs scripts topological order by default 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Test Workspace A
➤ YN0000: Test Workspace C
➤ YN0000: Test Workspace B
➤ YN0000: Done
",
}
`;

exports[`Commands workspace foreach shows prefix with flag 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: [workspace-a]: Test Workspace A
➤ YN0000: [workspace-c]: Test Workspace C
➤ YN0000: [workspace-b]: Test Workspace B
➤ YN0000: Done
",
}
`;
